File: todolist-python/app/utils/config_util.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
配置管理模块

提供统一的配置管理功能，集中存储和管理项目的配置信息。
"""

import os
import json
from .logger_util import logger


class ConfigUtil:
    """
    配置管理工具类
    """
    
    def __init__(self):
        """
        初始化配置管理工具
        """
        # 项目根目录
        self.project_root = os.path.join(os.path.dirname(os.path.dirname(os.path.dirname(__file__))))
        # 数据目录 - 使用Python项目自己的目录
        self.data_dir = os.path.join(self.project_root, 'data')
        # 缓存目录
        self.cache_dir = os.path.join(self.data_dir, 'cache')
        # 配置目录
        self.config_dir = os.path.join(self.data_dir, 'config')
        # 天气数据目录
        self.weather_dir = os.path.join(self.data_dir, 'weather')
        # 模拟数据目录 - 使用Python项目自己的mock目录
        self.mock_dir = os.path.join(self.data_dir, 'mock')
        
        # 配置定义，类似于Node项目中的CONFIG_DEFINITIONS
        self.config_definitions = {
            'app': {
                'file': 'app_config.json'
            },
            'notify': {
                'file': 'notify_config.json'
            },
            'festival': {
                'file': 'festival_config.json'
            }
        }
        
        # 确保所有目录存在
        os.makedirs(self.data_dir, exist_ok=True)
        os.makedirs(self.cache_dir, exist_ok=True)
        os.makedirs(self.config_dir, exist_ok=True)
        os.makedirs(self.weather_dir, exist_ok=True)
        os.makedirs(self.mock_dir, exist_ok=True)
    
    def get_project_root(self):
        """
        获取项目根目录
        
        Returns:
            项目根目录路径
        """
        return self.project_root
    
    def get_data_dir(self):
        """
        获取数据目录
        
        Returns:
            数据目录路径
        """
        return self.data_dir
    
    def get_cache_dir(self):
        """
        获取缓存目录
        
        Returns:
            缓存目录路径
        """
        return self.cache_dir
    
    def get_config_dir(self):
        """
        获取配置目录
        
        Returns:
            配置目录路径
        """
        return self.config_dir
    
    def get_weather_dir(self):
        """
        获取天气数据目录
        
        Returns:
            天气数据目录路径
        """
        return self.weather_dir
    
    def get_mock_dir(self):
        """
        获取模拟数据目录
        
        Returns:
            模拟数据目录路径
        """
        return self.mock_dir
    
    def _get_config_path(self, config_name):
        """
        根据配置名称获取配置文件的路径
        
        Args:
            config_name: 配置名称
        
        Returns:
            配置文件路径
        
        Raises:
            ValueError: 如果配置名称不存在
        """
        definition = self.config_definitions.get(config_name)
        if not definition:
            raise ValueError(f"Unknown config: {config_name}")
        return os.path.join(self.config_dir, definition['file'])
    
    def _read_config_file(self, config_path):
        """
        读取配置文件
        
        Args:
            config_path: 配置文件路径
        
        Returns:
            配置数据
        """
        try:
            if os.path.exists(config_path):
                with open(config_path, 'r', encoding='utf-8') as f:
                    return json.load(f)
            logger.warning(f"Config file not found: {config_path}")
            return {}
        except Exception as e:
            logger.error(f"读取配置文件失败: {str(e)}")
            return {}
    
    def get_config(self, config_name, default=None):
        """
        从配置文件中读取配置
        
        Args:
            config_name: 配置文件名称（不含扩展名）
            default: 默认值
        
        Returns:
            配置数据
        """
        try:
            config_path = self._get_config_path(config_name)
            config = self._read_config_file(config_path)
            return config or default
        except Exception as e:
            logger.error(f"获取配置失败: {str(e)}")
            return default
    
    def get_config_sync(self, config_name, default=None):
        """
        同步从配置文件中读取配置
        
        Args:
            config_name: 配置文件名称（不含扩展名）
            default: 默认值
        
        Returns:
            配置数据
        """
        return self.get_config(config_name, default)
    
    def save_config(self, config_name, config_data):
        """
        保存配置到配置文件
        
        Args:
            config_name: 配置文件名称（不含扩展名）
            config_data: 配置数据
        
        Returns:
            是否保存成功
        """
        try:
            config_path = self._get_config_path(config_name)
            
            # 确保配置目录存在
            config_dir = os.path.dirname(config_path)
            os.makedirs(config_dir, exist_ok=True)
            
            with open(config_path, 'w', encoding='utf-8') as f:
                json.dump(config_data, f, ensure_ascii=False, indent=2)
            
            logger.info(f"Config saved successfully: {config_name}")
            return True
        except Exception as e:
            logger.error(f"保存配置文件失败: {str(e)}")
            return False
    
    def get_config_value(self, config_name, key_path, default=None):
        """
        根据键路径获取配置值
        
        Args:
            config_name: 配置文件名称（不含扩展名）
            key_path: 键路径，如 'server.ports.python'
            default: 默认值
        
        Returns:
            配置值
        """
        try:
            config = self.get_config_sync(config_name)
            keys = key_path.split('.')
            value = config
            
            for key in keys:
                if value is None:
                    return default
                value = value.get(key)
            
            return value if value is not None else default
        except Exception as e:
            logger.error(f"获取配置值失败: {str(e)}")
            return default
    
    def get_available_configs(self):
        """
        获取所有可用的配置名称
        
        Returns:
            配置名称列表
        """
        return list(self.config_definitions.keys())
    
    def register_config(self, config_name, definition):
        """
        注册新的配置
        
        Args:
            config_name: 配置名称
            definition: 配置定义，包含 'file' 键
        
        Raises:
            ValueError: 如果配置名称已存在或定义不完整
        """
        if config_name in self.config_definitions:
            raise ValueError(f"Config {config_name} already registered")
        
        if 'file' not in definition:
            raise ValueError('Config definition must include file')
        
        self.config_definitions[config_name] = {
            'file': definition['file']
        }


# 创建全局配置工具实例
config_util = ConfigUtil()

File: todolist-go/app/utils/config_util.go
package utils

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"
)

// ConfigUtil 配置管理工具
type ConfigUtil struct {
	projectRoot       string
	dataDir           string
	cacheDir          string
	configDir         string
	weatherDir        string
	mockDir           string
	configDefinitions map[string]configDefinition
	configCache       map[string]map[string]interface{}
	configCacheMutex  sync.RWMutex
}

// configDefinition 配置定义
type configDefinition struct {
	File string `json:"file"`
}

// NewConfigUtil 创建配置管理工具实例
func NewConfigUtil() *ConfigUtil {
	projectRoot := getProjectRoot()
	dataDir := filepath.Join(projectRoot, "data")
	dirs := []string{dataDir, filepath.Join(dataDir, "cache"), filepath.Join(dataDir, "config"),
		filepath.Join(dataDir, "weather"), filepath.Join(dataDir, "mock")}
	for _, dir := range dirs {
		os.MkdirAll(dir, 0755)
	}
	return &ConfigUtil{
		projectRoot: projectRoot, dataDir: dataDir,
		cacheDir: filepath.Join(dataDir, "cache"), configDir: filepath.Join(dataDir, "config"),
		weatherDir: filepath.Join(dataDir, "weather"), mockDir: filepath.Join(dataDir, "mock"),
		configDefinitions: map[string]configDefinition{
			"app": {File: "app_config.json"}, "notify": {File: "notify_config.json"},
			"festival": {File: "festival_config.json"},
		},
		configCache: make(map[string]map[string]interface{}),
	}
}

// getProjectRoot 获取项目根目录
func getProjectRoot() string {
	// 首先检查是否在.app包中运行（通过可执行文件路径）
	exe, err := os.Executable()
	if err == nil {
		exeDir := filepath.Dir(exe)

		if strings.Contains(exeDir, ".app/Contents/MacOS") {
			// 在.app包中，返回Resources目录
			// 正确的路径应该是: .app/Contents/Resources
			resourcesDir := filepath.Join(filepath.Dir(exeDir), "Resources")
			if _, err := os.Stat(resourcesDir); err == nil {
				return resourcesDir
			}
		}
	}

	// 非.app包环境，按原逻辑查找
	currentDir, _ := os.Getwd()
	for {
		if _, err := os.Stat(filepath.Join(currentDir, "go.mod")); err == nil {
			return currentDir
		}
		parentDir := filepath.Dir(currentDir)
		if parentDir == currentDir {
			// 到达根目录
			return "."
		}
		currentDir = parentDir
	}
}

// Getter方法
func (c *ConfigUtil) GetProjectRoot() string { return c.projectRoot }
func (c *ConfigUtil) GetDataDir() string     { return c.dataDir }
func (c *ConfigUtil) GetCacheDir() string    { return c.cacheDir }
func (c *ConfigUtil) GetConfigDir() string   { return c.configDir }
func (c *ConfigUtil) GetWeatherDir() string  { return c.weatherDir }
func (c *ConfigUtil) GetMockDir() string     { return c.mockDir }

// getConfigPath 根据配置名称获取配置文件路径
func (c *ConfigUtil) getConfigPath(configName string) (string, error) {
	definition, ok := c.configDefinitions[configName]
	if !ok {
		return "", fmt.Errorf("unknown config: %s", configName)
	}
	return filepath.Join(c.configDir, definition.File), nil
}

// readConfigFile 读取配置文件
func (c *ConfigUtil) readConfigFile(configPath string) (map[string]interface{}, error) {
	data, err := os.ReadFile(configPath)
	if err != nil {
		if os.IsNotExist(err) {
			LoggerInstance.Warn("配置文件不存在", "config_path", configPath)
			return map[string]interface{}{}, nil
		}
		LoggerInstance.Error("读取配置文件失败", "error", err)
		return nil, err
	}

	var config map[string]interface{}
	if err := json.Unmarshal(data, &config); err != nil {
		LoggerInstance.Error("解析配置文件失败", "config_path", configPath, "error", err)
		// 与Python版本保持一致：解析失败时返回空字典，不返回错误
		return map[string]interface{}{}, nil
	}

	return config, nil
}

// GetConfig 从配置文件中读取配置
func (c *ConfigUtil) GetConfig(configName string, defaultConfig map[string]interface{}) map[string]interface{} {
	c.configCacheMutex.RLock()
	if cachedConfig, exists := c.configCache[configName]; exists {
		c.configCacheMutex.RUnlock()
		if len(cachedConfig) == 0 {
			return defaultConfig
		}
		return cachedConfig
	}
	c.configCacheMutex.RUnlock()

	configPath, err := c.getConfigPath(configName)
	if err != nil {
		LoggerInstance.Error("获取配置路径失败", "config_name", configName)
		return defaultConfig
	}
	config, _ := c.readConfigFile(configPath)

	c.configCacheMutex.Lock()
	c.configCache[configName] = config
	c.configCacheMutex.Unlock()

	if len(config) == 0 {
		return defaultConfig
	}
	return config
}

// GetConfigSync 同步从配置文件中读取配置
func (c *ConfigUtil) GetConfigSync(configName string, defaultConfig map[string]interface{}) map[string]interface{} {
	return c.GetConfig(configName, defaultConfig)
}

// SaveConfig 保存配置到配置文件
func (c *ConfigUtil) SaveConfig(configName string, configData map[string]interface{}) bool {
	configPath, err := c.getConfigPath(configName)
	if err != nil {
		LoggerInstance.Error("获取配置路径失败", "config_name", configName)
		return false
	}
	os.MkdirAll(filepath.Dir(configPath), 0755)
	data, _ := json.MarshalIndent(configData, "", "  ")
	if err := os.WriteFile(configPath, data, 0644); err != nil {
		LoggerInstance.Error("写入配置文件失败", "config_path", configPath)
		return false
	}

	c.configCacheMutex.Lock()
	c.configCache[configName] = configData
	c.configCacheMutex.Unlock()

	LoggerInstance.Info("配置保存成功", "config_name", configName)
	return true
}

// GetConfigValue 根据键路径获取配置值
func (c *ConfigUtil) GetConfigValue(configName string, keyPath string, defaultValue interface{}) interface{} {
	config := c.GetConfigSync(configName, map[string]interface{}{})
	if len(config) == 0 {
		return defaultValue
	}
	var value interface{} = config
	for _, key := range strings.Split(keyPath, ".") {
		if m, ok := value.(map[string]interface{}); ok {
			if v, exists := m[key]; exists {
				value = v
			} else {
				return defaultValue
			}
		} else {
			return defaultValue
		}
	}
	return value
}

// GetAvailableConfigs 获取所有可用的配置名称
func (c *ConfigUtil) GetAvailableConfigs() []string {
	configs := make([]string, 0, len(c.configDefinitions))
	// 按照固定顺序返回，与Python版本保持一致
	for _, configName := range []string{"app", "notify", "festival"} {
		if _, exists := c.configDefinitions[configName]; exists {
			configs = append(configs, configName)
		}
	}
	// 添加其他动态注册的配置
	for config := range c.configDefinitions {
		found := false
		for _, existing := range configs {
			if existing == config {
				found = true
				break
			}
		}
		if !found {
			configs = append(configs, config)
		}
	}
	return configs
}

// RegisterConfig 注册新的配置
func (c *ConfigUtil) RegisterConfig(configName string, definition configDefinition) error {
	if _, exists := c.configDefinitions[configName]; exists {
		return fmt.Errorf("config %s already registered", configName)
	}

	if definition.File == "" {
		return fmt.Errorf("config definition must include file")
	}

	c.configDefinitions[configName] = definition
	return nil
}

// 全局配置工具实例
var ConfigUtilInstance *ConfigUtil

// init 初始化全局配置工具实例
func init() {
	ConfigUtilInstance = NewConfigUtil()
}

// getAPITimeout 获取API超时时间（秒）
func (c *ConfigUtil) getAPITimeout(configKey string, defaultValue int) int {
	if t := c.GetConfigValue("app", configKey, nil); t != nil {
		if tFloat, ok := t.(float64); ok {
			return int(tFloat)
		}
	}
	return defaultValue
}

// GetLocationAPITimeoutDuration 获取位置API超时时间（time.Duration）
func (c *ConfigUtil) GetLocationAPITimeoutDuration() time.Duration {
	return time.Duration(c.getAPITimeout("server.api_timeout_location", 10)) * time.Second
}

// GetWeatherAPITimeoutDuration 获取天气API超时时间（time.Duration）
func (c *ConfigUtil) GetWeatherAPITimeoutDuration() time.Duration {
	return time.Duration(c.getAPITimeout("server.api_timeout_weather", 10)) * time.Second
}

// GetWeatherAPIuTLSTimeoutDuration 获取天气API uTLS超时时间（time.Duration）
func (c *ConfigUtil) GetWeatherAPIuTLSTimeoutDuration() time.Duration {
	return time.Duration(c.getAPITimeout("server.api_timeout_weather_utls", 20)) * time.Second
}
